[docs] Update sharktuner documentation (#20704)
openxla/iree: docs/website/docs/reference/tuning.md

---
hide:
  - tags
tags:
  - performance
icon: octicons/meter-16
status: new
---

# :octicons-meter-16: Tuning

Tuning is an important step in the process of extracting performance from your
hardware. During the compilation of a workload, IREE makes decisions on how to
set certain parameters that define how a workload is run on targeted hardware.
For example, when targeting a GPU, there could be multiple options for thread
count or tile size for a given input graph. By default, these parameters are
chosen in a way that performs well for any generic workload. However, it is
often possible to select values for these parameters that squeeze out extra
performance for a specific workload.

This process of iterating over the space of possible parameter values (knobs)
to enable improvements in some chosen performance metrics is called Tuning.

```mermaid
---
title: Generic Tuning workflow
---
graph LR;
    A[Compile]-->B;
    B[Benchmark]-->C;
    C[Set knobs]-->A;

> ```

Reviewer: Does this title render correctly on the website? You can also add an accessibility title and description, like on https://github.com/iree-org/iree/blob/1c075e0ab9b2ea640522bf3e673bf1d3e9068374/docs/website/docs/guides/ml-frameworks/pytorch.md?plain=1#L25-L35

The accessibility title shows up on hover, and screen readers should read back the description.

Author: It rendered how i expected when using `mkdocs serve`.

Also would you prefer I add an accessibility title and description to the other graphs too?

Reviewer: > Also would you prefer I add an accessibility title and description to the other graphs too?

That would be great.

## :octicons-dependabot-16: SHARK Tuner

### :octicons-book-16: Overview

> While tuning can be done manually, IREE provides support for a tool called
> [SHARK Tuner](https://github.com/nod-ai/shark-ai/tree/main/sharktuner) which is
> an automated dispatch tuner that can help speed up this process. To elaborate,
> an automated tuner can perform the iteration over possible values for knobs
> automatically. SHARK Tuner being an automated dispatch tuner, performs this
> automated iteration on dispatches instead of the entire model. To briefly
> describe a dispatch, dispatches are blocks of code that are created as part of
> IREE's compilation flow by splitting the input graph into blocks that can be
> executed concurrently and atomically. For further information on dispatches see
> the sections below.
> 
> So SHARK Tuner allows you to automatically tune individual dispatches in your
> model to achieve greater performance.

Reviewer: This section is quite verbose and repeats itself several times.

```suggestion
While tuning can be done manually, the
[SHARK Tuner](https://github.com/nod-ai/shark-ai/tree/main/sharktuner) tool
can automatically search through possible knob values for individual
dispatches to improve overall program performance. Dispatches are blocks of
code that are created as part of IREE's compilation flow by splitting the
input program into blocks that can be executed concurrently and atomically.
For further information on dispatches see the sections below.
```

Author: done

!!! info
    For more information about SHARK Tuner, see its source in the
    [shark-ai GitHub repository](https://github.com/nod-ai/shark-ai/tree/main/sharktuner)
    and the [Model Tuner example](https://github.com/nod-ai/shark-ai/tree/main/sharktuner/model_tuner).

In our experience, using the SHARK Tuner can provide **meaningful speedup** of
model execution.

!!! tip
    SHARK Tuner achieved a ~10% improvement on the SDXL

>     (Stable Diffusion XL) model with the MI300X GPU.

Reviewer: Could link to https://rocm.blogs.amd.com/artificial-intelligence/mi325x-accelerates-mlperf-inference/README.html#tuning here

Author: done

        %3 = tensor.empty() : tensor<32xf32>
        %4 = linalg.fill ins(%cst : f32) outs(%3 : tensor<32xf32>) -> tensor<32xf32>
        %5 = linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>, affine_map<(d0, d1) -> (d0)>], iterator_types = ["parallel", "reduction"]} ins(%2 : tensor<32x2048xf32>) outs(%4 : tensor<32xf32>) {
        ^bb0(%in: f32, %out: f32):
          %6 = arith.addf %in, %out : f32
          linalg.yield %6 : f32
        } -> tensor<32xf32>
        flow.dispatch.tensor.store %5, %1, offsets = [0], sizes = [32], strides = [1] : tensor<32xf32> -> !flow.dispatch.tensor<writeonly:tensor<32xf32>>
        return
      }
    }
  }
}
```

Illustrated graphically, the following dispatches were created.

```mermaid
flowchart TD
    A[/Input 1/]-->|tensor<32x1024xf32>|C;
    B[/Input 2/]-->|tensor<1024x2048xf32>|C;
    subgraph Dispatch 0
    C(MatMul);
    end
    C-->|tensor<32x2048xf32>|D
    subgraph Dispatch 1
    D(Sum Reduction);
    end
    D-->|tensor<32xf32>|F[/Output/];
```

In the above example, each `hal.executable` represents a dispatch.

> To oversimplify, when creating dispatches we first identify "root" operations,

Reviewer: nit: tone with "to oversimplify" feels strange. Reword?

Author: Reworded but ended up being a little more verbose, let me know how it reads to you

dispatches.

### :octicons-gear-16: Knobs in Dispatches

Depending on the hardware being targeted, a dispatch will expose different
knobs. Focusing on GPUs, some common knobs are subgroup tile sizes or workgroup
thread count. For a given input graph, the knobs associated with a particular
dispatch can be seen by adding the following flags when compiling.

* `--iree-hal-dump-executable-benchmarks-to=<directory>`
* `--iree-config-add-tuner-attributes`

These will dump standalone hal.executable benchmarks for each dispatch. Within
these benchmark files we can find an attribute associated to the root op of the
dispatch which shows some tunable attributes.

Shown below is the aforementioned attribute for the MatMul dispatch from our
earlier example.

```mlir
linalg.matmul {lowering_config = #iree_gpu.lowering_config<{mma_kind = #iree_gpu.mma_layout<MFMA_F32_16x16x16_F16>, promote_operands = [0, 1], reduction = [0, 0, 128], subgroup_m_count = 1 : i64, subgroup_n_count = 4 : i64, workgroup = [16, 128, 0]}>, root_op} ins(%3, %4 : tensor<32x1024xf16>, tensor<1024x2048xf16>) outs(%6 : tensor<32x2048xf32>) -> tensor<32x2048xf32>

> ```

Reviewer: Can this code block be wrapped at 80 characters or converted form a fenced code block with triple backticks to an inline code block with single backticks? It looks like it will overflow and have a very wide horizontal scroll bar.

Author: done

  %red = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel", "reduction"]} ins(%mm : tensor<32x2048xf32>) outs(%red_fill : tensor<32xf32>)  {
  ^bb0(%in: f32, %out: f32):
    %7 = arith.addf %in, %out : f32
    linalg.yield %7 : f32
  } -> tensor<32xf32>
  return %red: tensor<32xf32>
}
```

The above IR performs a MatMul and then a sum over the last dimension.

```mermaid
flowchart TD
    accTitle: Illustration of example IR
    accDescr {
     An example of IR representing a simple model consisting of a matmul with
     shape `32x1024x2048` followed by a reduction sum over the last dimension.
    }
    A[/Input 1/]-->|tensor<32x1024xf32>|C;
    B[/Input 2/]-->|tensor<1024x2048xf32>|C;
    C(MatMul)-->|tensor<32x2048xf32>|E;
    E(Sum Reduction)-->|tensor<32xf32>|F[/Output/];
```

While compiling this graph with IREE, the flag
`--iree-hal-dump-executable-files-to=<some directory>` can be used to observe
the created dispatches.

```mlir
// RUN: iree-compile --iree-hal-target-device=hip --iree-hip-target=gfx942 \
            --mlir-print-ir-after=iree-codegen-materialize-user-configs \

>             --iree-hal-dump-executable-files-to=<some directory>

Reviewer: As code samples get large and have the potential to drift from the spec / syntax, you could consider moving these into lit tests and then embedding those files into the website: https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#embedding-external-files . We have a few examples of that in the docs/website/ folder already.

Author: I'm not entirely sure if these are appropriate as tests since the lowering config attributes here may be expected to change as we continue working on IREE, so having this as a point of failure may only add unnecessary upkeep. But I could add do this in a follow up PR.

---
hide:
  - tags
tags:
  - sharktuner
  - SHARKTUNER

>   - Tuner

Reviewer: Pick 1 tag or none, not 3. Given that there will likely be only 1-2 pages for the tuner, I would vote to instead tag this "performance" or "optimization" then add that tag to other pages that have a similar focus (benchmarking guides, optimization options pages, etc.).

See https://iree.dev/community/tags/ for all current tags.

---

# Tuning

> Tuning is an important step in the process of optimizing performance for your

Reviewer 2: This text is very high-level and not accurate. 

Check the tuning section in this blog: https://rocm.blogs.amd.com/artificial-intelligence/mi325x-accelerates-mlperf-inference/README.html#stable-diffusion-xl-sdxl-text-to-image-mlperf-inference-benchmark to see how to improve the existing documents.

Author: The tuning section you linked mentions three ideas:
- 1A) Tuners adjust knobs
- 1B) SHARK Tuner picks the values for the knobs based on execution time
- 1C) SHARK Tuner chooses the spec that has the best execution time for the dispatch

My paragraph has these four ideas:
- 2A) Tuners explore configurations for tunable knobs
- 2B) SHARK Tuner improves performance by finding better specs for knobs
- 2C) Subgroup size and tiling sizes are examples of tunable knobs
- 2D) dispatches are subdivisions of a program, and a dispatch tuner explores specs for a specific dispatch

So, if I understand correctly, you want me to add 1C) to the description. Is there anything you want me to remove or rephrase in 2A), 2B), 2C), or 2D)?

Reviewer 2: The current first sentence, "Tuning is an important step in the process of optimizing performance for your applications," is partially accurate but not particularly informative. A more precise statement should reflect the actual objective of tuning:

Tuning in IREE is used to generate transform dialect (TD) specs that guide the IREE compiler in selecting optimal values for configurable parameters during the compilation phase.

Reviewer: > Tuning in IREE is used to generate transform dialect (TD) specs that guide the IREE compiler in selecting optimal values for configurable parameters during the compilation phase.

I would say that transform dialect specs are an implementation detail that could be left out of the first sentence. Focus on the user-visible impact of tuning in the introduction - programmatically trying variations of a program and selecting the ones that perform most efficiently.

Tuning is an important step in the process of optimizing performance for your

> applications. The IREE compiler provides default tuning specs (specifications)

Reviewer: This should also define "dispatch tuning".

Some pages for inspiration:

* https://tvm.apache.org/docs/v0.9.0/how_to/tune_with_autotvm/tune_relay_mobile_gpu.html
* https://dl.acm.org/doi/10.1145/3589236.3589241

Author: I chose to define it in the `# Overview` section, to not clutter the introduction. Is that okay?

Reviewer: I think it's pretty important to lead with it. If someone lands on this page without already knowing what a "dispatch" is or what "tuning" does, the first few lines of the document are the place to teach them or they will be confused or bounce to another page.

The IREE compiler is shipped with target-specific, default specs. With these
default specs, we aim to cover the most in-demand hardware and dispatches from
the most popular ML models, although we do not guarantee completeness. So for
many users, the default specs will provide sufficient performance, but are far
from perfect.

For users looking to get the best performance on custom models and hardware
targets, the IREE compiler also supports user-provided tuning specs. These
specs can be made manually or generated using our dispatch tuner, [SHARK Tuner](https://github.com/nod-ai/shark-ai/tree/main/sharktuner).

> A dispatch tuner is simply a tuner which performs tuning on units of 

Reviewer: trim trailing space to appease the pre-commit check: https://github.com/iree-org/iree/actions/runs/14867173837/job/41746733185?pr=20704#step:4:183
```suggestion
A dispatch tuner is simply a tuner which performs tuning on units of
```

> A dispatch tuner is simply a tuner which performs tuning on units of compute

Reviewer 2: compute -> computation?

> A dispatch tuner is simply a tuner which performs tuning on units of compute

Reviewer 2: Also here, computation

> A dispatch tuner is simply a tuner which performs tuning on units of compute

Reviewer 2: You can refer reader to this link: https://github.com/iree-org/iree/blob/main/samples/custom_dispatch/README.md for more detailed about "Dispatches" in IREE.

> compuitation called a "dispatch". Dispatches are blocks of computation

Reviewer: typo
```suggestion
computation called a "dispatch". Dispatches are blocks of computation
```

The IREE compiler is shipped with target-specific, default specs. With these
default specs, we aim to cover the most in-demand hardware and dispatches from
the most popular ML models, although we do not guarantee completeness. So for
many users, the default specs will provide sufficient performance.

> But for users looking to get the best performance on custom models and hardware
> targets, the IREE compiler also supports user-provided tuning specs. These

Reviewer: Don't start a sentence with "but"

```suggestion
For users looking to get the best performance on custom models and hardware
targets, the IREE compiler also supports user-provided tuning specs. These
```

specs can be made manually or generated using SHARK Tuner, which is still under
development.

> !!! info
>     Additional information on installation and a hands-on
>     [example](https://github.com/nod-ai/shark-ai/tree/main/sharktuner/model_tuner)
>     is available in an external
>     [repository](https://github.com/nod-ai/shark-ai/tree/main/sharktuner).

Reviewer: How about this rewording?

```suggestion
!!! info
    For more information about SHARK Tuner, see its source in the
    [shark-ai GitHub repository](https://github.com/nod-ai/shark-ai/tree/main/sharktuner)
    and the
    [Model Tuner example](https://github.com/nod-ai/shark-ai/tree/main/sharktuner/model_tuner).
```

* Lead with the general link (README, project overview) _then_ put the specific link (example package)
* https://developers.google.com/style/cross-references#descriptive-link-text

    > For the link text itself, use short, unique, descriptive phrases that provide context for the material that you're linking to.

best performance on custom models and hardware targets without having to modify
the compiler source code or needlessly special-case compiler heuristics.

This document describes how to work with tuning specs generated by the SHARK

> Tuner or produced manually, but it does not go into detail on how to generate

Reviewer: Let's structure this to more active voice that leads with the tuner package and then explains how the components fit together.

See https://iree.dev/guides/ml-frameworks/pytorch/ for an example page that has the structure I'd like to see:

* Link to the project
* Explanation of what it does
* Some diagrams and other formatting to break up the walls of text

Here is the current generated page for tuning.md: https://iree.dev/reference/tuning/ for comparison. Instructions for how to preview changes locally are at https://github.com/iree-org/iree/tree/main/docs/website#iree-user-facing-documentation-website.

Minimally, I would reorder this paragraph to something more like

```
The [SHARK Tuner](https://github.com/nod-ai/shark-ai/tree/main/sharktuner)
project provides one way to generate tuning specs and run them through
a tuning loop.
```

Ideally I would like to see some diagrams, installation instructions, and minimal code samples. We could also put an "info" or "note" admonition (https://squidfunk.github.io/mkdocs-material/reference/admonitions/) laying out the potential roadmap for SHARK tuner moving into IREE, or the possibility of other similar downstream producers of tuning specs.

these specs. Currently, the dispatch tuner that generates tuning specs is still

> experimental and hosted

Reviewer 3: ```suggestion
under development and hosted
```
I remember experimental was dropped in some other places thus we maybe shouldn't call it that way here I guess.

Reviewer: +1, don't call it experimental at this point. Could still mention status ("under development", "under active development", "evolving"), though such language tends to get outdated when written into docs :P

> [in an external repo](https://github.com/nod-ai/shark-ai/tree/main/sharktuner),

Reviewer 3: ```suggestion
in an [external repository](https://github.com/nod-ai/shark-ai/tree/main/sharktuner),
```

A dispatch tuner is simply a tuner which performs tuning on units of compute
called a "dispatch". Dispatches are blocks of compute generated by IREE to
allow a program to be divided into subdivisions capable of running

> concurrently. For more information on dispatches, see this [explanation](https://github.com/iree-org/iree/blob/main/samples/custom_dispatch/README.md).

Reviewer 2: Also use relative link?

Reviewer: No. When linking to source code from the website, use an absolute URL, not relative. When linking to another page _on the website_, use relative URLs.

Reviewer 2: But here this is a link to the readme.md. So, should we use a relative or an absolute URL?

Reviewer: Absolute. Only files under `docs/website/` get pages on the website. You can check the link behavior by building the website locally - I'm pretty sure a relative link like this will 404.

Reviewer 2: ok, got it. Please follow suggestions from Scott and ignore mine about this @Muzammiluddin-Syed-ECE.
